resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkhead.java

```diff
@@ -0,0 +1,546 @@
+		};
+	}
+
+	/**
+	 * Returns a function which is decorated by a bulkhead.
+	 *
+	 * @param bulkhead the bulkhead
+	 * @param function the original function
+	 * @param <T>      the type of the input to the function
+	 * @param <R>      the type of the result of the function
+	 * @return a function which is decorated by a bulkhead.
+	 */
+	static <T, R> CheckedFunction1<T, R> decorateCheckedFunction(AdaptiveBulkhead bulkhead, CheckedFunction1<T, R> function) {
+		return (T t) -> {
+			bulkhead.acquirePermission();
+			Instant start = null;
+			try {
+				start = Instant.now();
+				return function.apply(t);
+			} catch (Exception e) {
+				handleError(bulkhead, start, e);
+				throw e;
+			} finally {
+				if (start != null) {
+					Instant finish = Instant.now();
+					bulkhead.onComplete(Duration.between(start, finish), true);
+				}
+			}
+		};
+	}
+
+	static void handleError(AdaptiveBulkhead bulkhead, @Nullable Instant start, Exception e) {
+		if (bulkhead.getBulkheadConfig().getAdaptIfError() != null) {
+			//noinspection unchecked
+			if (bulkhead.getBulkheadConfig().getAdaptIfError().test(e) && start != null) {
+				Instant finish = Instant.now();
+				bulkhead.onComplete(Duration.between(start, finish), false);
+
+			}
+		} else {
+			bulkhead.releasePermission();
+		}
+	}
+
+	/**
+	 * Create a Bulkhead with a default configuration.
+	 *
+	 * @param name the name of the bulkhead
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead ofDefaults(String name) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, AdaptiveBulkheadConfig.ofDefaults());
+	}
+
+
+	/**
+	 * Creates a bulkhead with a custom configuration and custom limiter
+	 *
+	 * @param name         the name of the bulkhead
+	 * @param config       a custom BulkheadConfig configuration
+	 * @param limitAdapter the custom limit adopter
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, AdaptiveBulkheadConfig config, LimitPolicy<Bulkhead> limitAdapter) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, config, limitAdapter);
+	}
+
+	/**
+	 * Creates a bulkhead with a custom configuration
+	 *
+	 * @param name   the name of the bulkhead
+	 * @param config a custom BulkheadConfig configuration
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, AdaptiveBulkheadConfig config) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, config);
+	}
+
+	/**
+	 * Creates a bulkhead with a custom configuration
+	 *
+	 * @param name                   the name of the bulkhead
+	 * @param bulkheadConfigSupplier custom configuration supplier
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, Supplier<AdaptiveBulkheadConfig> bulkheadConfigSupplier) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, bulkheadConfigSupplier.get());
+	}
+
+	interface AdaptiveBulkheadMetrics extends Bulkhead.Metrics {
```

Comment:
We should call the interface `Metrics` as well. 
We always used `Bulkhead.Metrics`, `CircuitBreaker.Metrics`, `RateLimiter.Metrics`.

Can we also include `getNumberOfFailedCalls`,  `getNumberOfSuccessfulCalls`, and `getNumberOfSlowCalls`. 
I think we can should getAverageLatencyMillis.